### ISSUE TITLE
WebSocket connection failed errors after upgrade to Next.js 12

### ISSUE THREAD

**Issue by matt-joecoffee** [vercel/next.js#30491]
### What version of Next.js are you using?

12.0.1

### What version of Node.js are you using?

12.22.3

### What browser are you using?

Brave

### What operating system are you using?

macOS

### How are you deploying your application?

next start

### Describe the Bug

After upgrading to Next.js 12, I'm getting the following error while running the app in `dev` mode:

```
WebSocket connection to 'wss://app.name.dev/_next/webpack-hmr' failed: 
```

We're using nginx to proxy the domain, so that we actually have a domain name, rather than just using localhost. I can't think of any other gotchas that might be contributing factors.

The error repeats every few seconds, on a regular interval, and eventually floods the browser console. I am not seeing any errors printed to the terminal.

### Expected Behavior

No errors!

### To Reproduce

n/a

**Comment by pmoskvin** [comment]
I have the same problem. Nginx is not used

**Comment by KillerCodeMonkey** [comment]
same here, just upgraded to v12.0.1 from v11 (without touching anything than increasing the nextjs version in the package.json). running dev server works, but hmr seems not working.

Not blocking since i can just manually reload, but it would be nice to get it back working :).

Thanks for your great work

**Comment by cabbiepete** [comment]
FWIW I was also having this issue and @ijjk's comment (https://github.com/vercel/next.js/issues/30491#issuecomment-953337935) put me in the correct direction. Previously there had not been a need to include the web sockets proxying for HMR to work but it seems now there is. After I added to proxying to the Next.JS server and restarted Nginx I no longer have this issue.

@matt-joecoffee We are also serving the site from a real domain in a similar way I suspect to you.

```
# Websockets
proxy_http_version 1.1;
proxy_set_header Upgrade $http_upgrade;
proxy_set_header Connection "upgrade";
```

**Comment by timneutkens** [comment]
@ijjk we'll want to add this to the upgrading guide 👍

**Comment by Manc** [comment]
Sorry for coming back to this, but can anybody help find a solution when using a custom Express server?

As I understand, the problem here is slightly different, because the WS request cannot be forwarded/proxied to a HTTP port, because Next.js handles requests using a JavaScript function.

Example (abbreviated):

```TypeScript
import createNextServer from 'next';
const nextServer = createNextServer(/*...*/);
const nextRequestHandler = nextServer.getRequestHandler();
nextServer
	.prepare()
	.then(() => {
		const server = express();
		/* Other routes… */
		server.all('*', (req, res) => nextRequestHandler(req, res));
	});
```

**Comment by ijjk** [comment]
@Manc it should be working with the above case can you share the network error/log showing it not working?

**Comment by KillerCodeMonkey** [comment]
It's the same issue for me. I have a custom express server.

The error in the js console of the browser is that a ws connection to
_next/webpack-hmr can not be established (I am in my handy currently,
sorry).

I tried adding a proxy as Express Middleware for the giving url to Upgrade
ws connections. But it looks like the webpack hmr is not running/reachable
at all?



JJ Kasper ***@***.***> schrieb am Mo., 1. Nov. 2021, 17:22:

**Comment by marnixhoh** [comment]
Hi,
I am having the same issue on my local machine. I am using a custom server (only in local development) to proxy the socket requests. I am not using Nginx.

I am getting the following error in my console where Next.js is running:
`Error occurred while proxying request www.XXX.com.test:3001/_next/webpack-hmr to undefined`

And in the browser console:
`WebSocket connection to 'ws://www.XXX.com.test:3001/_next/webpack-hmr' failed: One or more reserved bits are on: reserved1 = 1, reserved2 = 0, reserved3 = 0`

In case it helps, here is the custom server code:
```
const express = require('express')
const next = require('next')
const { createProxyMiddleware } = require('http-proxy-middleware')

const port = 3001
const app = next({ dev: true })
const handle = app.getRequestHandler()

app.prepare().then(() => {
    const server = express()

    server.use('/api', createProxyMiddleware({ target: 'http://localhost:3000', ws: true }))

    server.all('*', (req, res) => {
        return handle(req, res)
    })

    server.listen(port, (err) => {
        if (err) throw err
        console.log(`> Ready on http://localhost:${port}`)
    })
})
```

**Comment by fabb** [comment]
Same issue here, not using nginx, but a custom express server similar to the one above. 

I get a bit different error though:
```
WebSocket connection to 'ws://localhost:3010/_next/webpack-hmr' failed: WebSocket is closed before the connection is established.
```

Happy for any pointers.

**Comment by fabb** [comment]
Ok, i fixed it simply with this:
```
    server.all(/^\/_next\/webpack-hmr(\/.*)?/, async (req: Request, res: Response) => {
        void handle(req, res)
    })
```

be aware of https://github.com/vercel/next.js/issues/18080, if you use a custom `assetPrefix` in dev mode, you need another middleware to rewrite the path:


```
    server.use(function (req: Request, _res: Response, nextFn: NextFunction) {
        if (!req.originalUrl.includes('webpack-hmr')) {
            return nextFn()
        }
        req.url = req.url.replace('/my-prefix/_next/webpack-hmr', '/_next/webpack-hmr')
        return nextFn()
    })

    server.all(/^\/_next\/webpack-hmr(\/.*)?/, async (req: Request, res: Response) => {
        void handle(req, res)
    })
```

**Comment by skmohammadi** [comment]
For me, this code worked:
```
location /_next/webpack-hmr {
	proxy_pass http://loadbalancer/_next/webpack-hmr;
	proxy_http_version 1.1;
	proxy_set_header Upgrade $http_upgrade;
	proxy_set_header Connection "upgrade";
}
```

where **loadbalacer** is local docker server.

**Comment by KillerCodeMonkey** [comment]
hey there and the provided solution for routing all hmr requests to the nextjs handler is not working for me.
The cause for it is, that i have an apollo server with websockets as well.
So my custom server is defining a custom websocket route:

```TS
const wsServer = new ws.Server({
  server: httpServer,
  path: '/graphql'
})
```

Searching for a solution...

**Comment by lucaslcode** [comment]
Just to add, I was having this issue without any kind of middleware/proxy, I just restarted my computer to fix. Possible something WSL related.

**Comment by Manc** [comment]
I finally solved the issue I was having (see comment above). I use both a WS connection in my web app, which needs to be proxied in development, and Webpack HMR for development.

This is my modified custom server script (abbreviated):

```ts
import http from 'http';
import createNextServer from 'next';
import { createProxyMiddleware } from 'http-proxy-middleware';

const port = parseInt(process.env.PORT || '3000', 10);
const nextServer = createNextServer(/*...*/);
const nextRequestHandler = nextServer.getRequestHandler();

nextServer
	.prepare()
	.then(() => {
		const expressInstance = express();

		// Set up the WebSocket API proxy.
		expressInstance.use(
			createProxyMiddleware('/path/to/my/websocket/endpoint', {
				target: 'ws://192.168.0.123:1234',
				ws: true,
			})
		);

		// Set up REST API proxy.
		expressInstance.use('/api', createProxyMiddleware({
			target: 'http://192.168.0.123:1234',
		});

		/* Other routes… */
		server.all('*', (req, res) => nextRequestHandler(req, res));

		http.createServer(expressInstance).listen(port);
	});
```

**Comment by Mahdi-Soultana** [comment]
see here
websocket.js?a9be:38 WebSocket connection to 'ws://localhost:3000/_next/webpack-hmr' failed: WebSocket is closed before the connection is established.

**Comment by focux** [comment]
I'm still experiencing this exact issue, which I think is related to having a custom server + `http-proxy-middleware` with WebSockets enabled. If I disable WebSocket on the proxy middleware, it starts working again. This is how my custom server looks:
```typescript
import { getAccessToken, getSession } from "@auth0/nextjs-auth0";
import { loadEnvConfig } from "@next/env";
import express from "express";
import { createProxyMiddleware } from "http-proxy-middleware";
import next from "next";

loadEnvConfig("./", process.env.NODE_ENV !== "production");

const port = parseInt(process.env.PORT, 10) || 3000;
const dev = process.env.NODE_ENV !== "production";
const app = next({ dev });
const handle = app.getRequestHandler();

app
  .prepare()
  .then(() => {
    const server = express();

    server.use("/graphql", async (req, res, next) => {
      try {
        const { accessToken } = await getAccessToken(req, res, { refresh: true });

        if (accessToken) {
          req.headers["Authorization"] = accessToken;
        }
      } catch (err) {
        res.json({ errors: [{ code: err.code, message: err.message }] });
      }
      next();
    });

    server.use(
      "/graphql",
      createProxyMiddleware({
        target: process.env.NEXT_PUBLIC_BACKEND_URL,
        changeOrigin: true,
        ws: true,
        onProxyReqWs: async (proxyReq, req, socket, server, head) => {
          if (req.url !== "/graphql") {
            return;
          }

          socket.on("error", function (error) {
            console.warn("Websockets error.", error, req.url);
          });

          try {
            const session = getSession(req, {} as any);
            if (session?.accessToken) {
              proxyReq.setHeader("Authorization", session.accessToken);
            }
          } catch (err) {}
        },
      })
    );

    server.all("*", (req, res) => {
      return handle(req, res);
    });

    server.listen(port, () => {
      console.log(`> Ready on http://localhost:${port}`);
    });
  })
  .catch((err) => console.log("error", err));

```

**Comment by janaroj** [comment]
Having a similar issue, after adding the aforementioned nginx config the request has the headers Connection: Upgrade, Upgrade: websocket but getting an error "WebSocket connection to '<URL>' failed: WebSocket is closed before the connection is established." - has anyone had a similar issue?

**Comment by NERLOE** [comment]
Does anyone have a solution? I still have the same problem.

**Comment by zestzero** [comment]
If you're using Nextjs with Nestjs, it might relate to this one https://stackoverflow.com/questions/67035266/socket-connection-from-nextjs-to-a-node-backend @Jay McDoniel mentioned that nest 8 will solve the issue.

**Comment by ccreeger** [comment]
We are using Charles Proxy for local development to map our URL (i.e. `https://www.EXAMPLE.test`) to `http://localhost:3000`. I was able to solve this problem by adding an additional Map Remote Settings for the web socket.

From: `wss://www.EXAMPLE.test`

To: `ws://localhost:3000`

**Comment by archy-bold** [comment]
@KillerCodeMonkey The only solution I found was to run a separate server for GraphQL from the next server ie a http server listening on a different port.

**Comment by richardpickett** [comment]
For those using apache, here are the proxy settings you'll need in your virtualhost, in this case I'm running node on port 10510, please adjust to suite your setup:

```
  ProxyPreserveHost On
  ProxyPassMatch ^/(_next/webpack-hmr)$  ws://localhost:10510/$1
  ProxyPass / http://localhost:10510/
  ProxyPassReverse / http://localhost:10510/
```

**Comment by thorep** [comment]
I have this problem on my nextjs12. Running locally..no custom server..

**Comment by sidd1707** [comment]
> For me, this code worked:
> 
> ```
> location /_next/webpack-hmr {
> 	proxy_pass http://loadbalancer/_next/webpack-hmr;
> 	proxy_http_version 1.1;
> 	proxy_set_header Upgrade $http_upgrade;
> 	proxy_set_header Connection "upgrade";
> }
> ```
> 
> where **loadbalacer** is local docker server.



> FWIW I was also having this issue and @ijjk's comment ([#30491 (comment)](https://github.com/vercel/next.js/issues/30491#issuecomment-953337935)) put me in the correct direction. Previously there had not been a need to include the web sockets proxying for HMR to work but it seems now there is. After I added to proxying to the Next.JS server and restarted Nginx I no longer have this issue.
> 
> @matt-joecoffee We are also serving the site from a real domain in a similar way I suspect to you.
> 
> ```
> # Websockets
> proxy_http_version 1.1;
> proxy_set_header Upgrade $http_upgrade;
> proxy_set_header Connection "upgrade";
> ```
> 
> Obviously not helpful for non nginx situations like @pmoskvin's but perhaps the package there is not forwarding these same headers and upgrading the http version?

Thanks for sharing. It worked in my case where i am using next js with fastify server and nginx for proxy

**Comment by jmitrikUSAP** [comment]
### I found a solution -- dumb as I may be ;)

Version Info:
Next version: 12.1.1-canary.7
Using a Custom Server with Express
Using PM2 to host from a windows server & IIS reverse proxy url rewrite

My fix:
In `package.json` scripts:

```
  "scripts": {
    "dev": "cross-env NODE_ENV=development node server.js",
    "build": "next build",
    "test": "cross-env NODE_ENV=test node server.js",
    "start": "cross-env NODE_ENV=production node server.js",
    "info": "next info"
  },
```
 In `server.js`
```
const dev = process.env.NODE_ENV !== 'production' 
const test = process.env.NODE_ENV === 'test'
const app = next({ dev })
```
I was running my "test" script on a test sever that is a pre-prod copy of the prod environment not realizing that 
`const app = next({ dev })` is what causes the wss to run , as the hmr is the way Next updates your modules on code changes. (**Yeah, that value right there is what controls the spinning little triangle at the bottom of your dev pages (the hot refresh indicator)**)

So even though everything else was identical between my test version and my production version,

### **_My test env Next Custom Express Server was evaluating to dev:true_** 

My fix:

 In `server.js` 
```
const app = next({ dev: process.env.NODE_ENV === 'development' })

const dev = process.env.NODE_ENV !== 'production' //test or development is true
const test = process.env.NODE_ENV === 'test' 
```
I ended up using the dev and test flags in my server file _a lot_ to give myself one codebase between my environments, so I removed them as a dependency on the fast refresh indication and only do fast refresh when I explicitly pass 'development' as my NODE_ENV.

Hope this helps someone. 


### tl:dr

If you have a Custom Express Server
Make sure  `const app = next({ dev })` dev evaluates to false in live or pseudo-live environments.

**Comment by jeremypress** [comment]
If anyone is running into this because they are using a custom next 12 server and apollo graphql subscriptions, we found the easiest workaround was to run the apollo sub server on a different port locally. Apollo + express are able to reconcile the new port, and HMR works as expected. Happy to share more if anyone runs into this.

**Comment by Kingwzb** [comment]
> If anyone is running into this because they are using a custom next 12 server and apollo graphql subscriptions, we found the easiest workaround was to run the apollo sub server on a different port locally. Apollo + express are able to reconcile the new port, and HMR works as expected. Happy to share more if anyone runs into this.

I ran into exactly the same issue with custom next server and apollo graphql subscriptions, do you mind sharing your solution with some code snippets?

**Comment by jeremypress** [comment]
> I ran into exactly the same issue with custom next server and apollo graphql subscriptions, do you mind sharing your solution with some code snippets?

Sure!

1. We didn't touch the next side, but make sure you are handling HMR requests like the example in https://github.com/vercel/next.js/issues/30491#issuecomment-956174988
2. I set a port and passed it into our apollo server config like this
```javascript
    const options = {
      server,
      path: '/graph-subscription-route',
    };

    if (process.env.NODE_ENV === 'development') {
      // Override to prevent conflicts with next12+ hmr
      options.port = process.env.DEV_GRAPHQL_SUBSCRIPTION_PORT;
    }


And then passed that options object to the apollo config


new SubscriptionServer(
      {
        execute,
        subscribe,
        schema,
        keepAlive: xxx,
        onDisconnect: (ws, context) => {
        ...
        },
        onConnect: async (connectionParams, ws) => {
        ...
      },
      options,


```


Hope this helps!

**Comment by Kingwzb** [comment]
> > I ran into exactly the same issue with custom next server and apollo graphql subscriptions, do you mind sharing your solution with some code snippets?
> 
> Sure!
> 
> 1. We didn't touch the next side, but make sure you are handling HMR requests like the example in [WebSocket connection failed errors after upgrade to Next.js 12 #30491 (comment)](https://github.com/vercel/next.js/issues/30491#issuecomment-956174988)
> 2. I set a port and passed it into our apollo server config like this
> 
> ```js
>     const options = {
>       server,
>       path: '/graph-subscription-route',
>     };
> 
>     if (process.env.NODE_ENV === 'development') {
>       // Override to prevent conflicts with next12+ hmr
>       options.port = process.env.DEV_GRAPHQL_SUBSCRIPTION_PORT;
>     }
> 
> 
> And then passed that options object to the apollo config
> 
> 
> new SubscriptionServer(
>       {
>         execute,
>         subscribe,
>         schema,
>         keepAlive: xxx,
>         onDisconnect: (ws, context) => {
>         ...
>         },
>         onConnect: async (connectionParams, ws) => {
>         ...
>       },
>       options,
> ```
> 
> Hope this helps!

Thanks for sharing it.

I tried but it seems SubscriptionServer code doesn't allow the passing server and port together, I got the below error at runtime.
error - unhandledRejection: TypeError: One and only one of the "port", "server", or "noServer" options must be specified

the package versions I used are
"subscriptions-transport-ws": "^0.11.0",
"next": "^12.1.0",

**Comment by Kingwzb** [comment]
With jeremypress's help, I finally made it working

server:
`console.log('node env : ', process.env.NODE_ENV);
const dev = process.env.NODE_ENV !== 'production';
const port = (process.env.PORT && parseInt(process.env.PORT, 10)) || 3080;
let subPort = port;
if(dev) {
    subPort = (process.env.SUB_PORT && parseInt(process.env.SUB_PORT, 10)) || 3081;
}

const nextServer = createNextServer({ dev });
const nextRequestHandler = nextServer.getRequestHandler();

nextServer.prepare().then(async () => {
    const app = express();
    //app.all('*', (req, res) => nextRequestHandler(req, res));
    const httpServer = http.createServer(app);
    const options: WebSocket.ServerOptions = {
        path: '/graphql',
    };
    if (dev) {
        // Override to prevent conflicts with next12+ hmr
        options.port = subPort;
    } else {
        options.server = httpServer;
    }
    console.log(JSON.stringify(options));
    const subscriptionServer = SubscriptionServer.create(
        {
            // This is the `schema` we just created.
            schema,
            // These are imported from `graphql`.
            execute,
            subscribe,
            // Providing `onConnect` is the `SubscriptionServer` equivalent to the
            // `context` function in `ApolloServer`. Please [see the docs](https://github.com/apollographql/subscriptions-transport-ws#constructoroptions-socketoptions--socketserver)
            // for more information on this hook.
            async onConnect(
                _connectionParams: Object,
                _webSocket: WebSocket,
                _context: ConnectionContext
            ) {
                // If an object is returned here, it will be passed as the `context`
                // argument to your subscription resolvers.
            },
        },
        options
    );

    const apolloServer = new ApolloServer({
        schema,
        context() {
            //loopup userId by token, etc.
        },
        plugins: [
            {
                async serverWillStart() {
                    return {
                        async drainServer() {
                            subscriptionServer.close();
                        },
                    };
                },
            },
        ],
    });
    await apolloServer.start();
    apolloServer.applyMiddleware({ app });
    app.use((req, res) => nextRequestHandler(req, res));
    //https://github.com/vercel/next.js/issues/30491
    //websocket.js?a9be:46  WebSocket connection to 'ws://localhost:3080/_next/webpack-hmr' failed
    httpServer.listen(port, () => {
        console.log('dev ', dev);
        console.log(`🚀 Http Server ready at http://localhost:${port}`);
        console.log(
            `🚀 GraphQL Server ready at http://localhost:${port}${apolloServer.graphqlPath}`
        );
        console.log(
            `🚀 GraphQL Subscriptions ready at ws://localhost:${subPort}${apolloServer.graphqlPath}`
        );
    });
});`

client:
`
let ssrMode = typeof window == 'undefined';
const dev = process.env.NODE_ENV !== 'production';
console.log('ssrMode : ' + ssrMode + ' dev: '+dev);
const port = (process.env.PORT && parseInt(process.env.PORT, 10)) || 3000;
let wsPort = port;
if(dev) {
    wsPort = (process.env.DEV_SUB_PORT && parseInt(process.env.DEV_SUB_PORT, 10)) || 3081;
}
let httpURI: string = '/graphql';
let wsURI = '';
if (ssrMode) {
    httpURI = `http://localhost:${port}/graphql`;
    wsURI = `ws://localhost:${wsPort}/graphql`;
} else {
    let isHttps = window.location.protocol == 'https:';
    wsURI =
        (isHttps ? 'wss' : 'ws') +
        '://' +
        window.location.hostname +
        ':' +
        wsPort +
        '/graphql';
}
`

**Comment by cyrusae** [comment]
I have the same problem (Next.js 12.1.4, Express server run on localhost:3001 with no intent of going to elsewhere ever) happening at random. For a while the error was happening but not preventing the app from working, now it kills some pages but not others, I added a component in one branch and now nothing in that branch loads. As far as I can tell it's not at all related to the code itself and everything I can find about it either says it's been fixed, has fixes for things I'm not using (e.g. nginx), or both.
```WebSocket connection to 'ws://localhost:3000/_next/webpack-hmr' failed: WebSocket is closed due to suspension.```

**Comment by fsolarv22** [comment]
> I have this problem on my nextjs12. Running locally..no custom server..

Same problem here, regular local server and no custom server (no express). I'm using windows 11 and Node v16

**Comment by Raghu-M-S** [comment]
> Hi, I am having the same issue on my local machine. I am using a custom server (only in local development) to proxy the socket requests. I am not using Nginx.
> 
> I am getting the following error in my console where Next.js is running: `Error occurred while proxying request www.XXX.com.test:3001/_next/webpack-hmr to undefined`
> 
> And in the browser console: `WebSocket connection to 'ws://www.XXX.com.test:3001/_next/webpack-hmr' failed: One or more reserved bits are on: reserved1 = 1, reserved2 = 0, reserved3 = 0`
> 
> In case it helps, here is the custom server code:
> 
> ```
> const express = require('express')
> const next = require('next')
> const { createProxyMiddleware } = require('http-proxy-middleware')
> 
> const port = 3001
> const app = next({ dev: true })
> const handle = app.getRequestHandler()
> 
> app.prepare().then(() => {
>     const server = express()
> 
>     server.use('/api', createProxyMiddleware({ target: 'http://localhost:3000', ws: true }))
> 
>     server.all('*', (req, res) => {
>         return handle(req, res)
>     })
> 
>     server.listen(port, (err) => {
>         if (err) throw err
>         console.log(`> Ready on http://localhost:${port}`)
>     })
> })
> ```

**Comment by Raghu-M-S** [comment]
where should i add the above server code in Nextjs

**Comment by archy-bold** [comment]
@Raghu-M-S It just needs to be in its own file eg `server.js`. Then update your scripts to run that file. [It's all in the documentation](https://nextjs.org/docs/advanced-features/custom-server).

**Comment by supercede** [comment]
I have this same issue but I'm not using any customer server. Running Next.js 12 on Windows 11 with wsl2
